core/src/main/kotlin/org/evomaster/core/search/gene/Gene.kt

```diff
@@ -79,45 +84,98 @@ abstract class Gene(var name: String) {
+            enableAdaptiveGeneMutation: Boolean = false,
+            additionalGeneMutationInfo: AdditionalGeneSelectionInfo? = null
+    ){
+        val internalGenes = candidatesInternalGenes(randomness, apc, allGenes, internalGeneSelectionStrategy, enableAdaptiveGeneMutation, additionalGeneMutationInfo)
+        if (internalGenes.isEmpty()){
+            val mutated = mutate(randomness, apc, mwc, allGenes, internalGeneSelectionStrategy, enableAdaptiveGeneMutation, additionalGeneMutationInfo)
+            if (!mutated) throw IllegalStateException("leaf mutation is not implemented")
+        }else{
+            val selected = selectSubset(internalGenes, randomness, apc, mwc, allGenes, internalGeneSelectionStrategy, enableAdaptiveGeneMutation, additionalGeneMutationInfo)
+            if (selected.isEmpty())
+                throw IllegalStateException("none is selected to mutate")
+
+            selected.forEach{
+                do {
+                    it.key.standardMutation(randomness, apc, mwc, allGenes, internalGeneSelectionStrategy, enableAdaptiveGeneMutation, it.value)
+                }while (!mutationCheck())
+            }
+        }
+    }
 
     /**
-     * Apply a archived-based mutation to the current gene.
-     *
-     * NOTE THAT if this method is not overridden, just default to standard mutation
-     *
-     *   @param randomness the source of non-determinism
-     *   @param allGenes if the gene depends on the other (eg a Foreign Key in SQL databases),
-     *          we need to refer to them
-     *   @param evi the evaluated individual contains an evolution of the gene with fitness values
-     *   @param selection how to select genes to mutate if [this] contains more than one genes(e.g., ObjectGene) or other characteristics(e.g., size of ArrayGene)
-     *   @param impact info of impact of the gene if it has, but in some case impact might be null, e.g., an element at ArrayGene
-     *   @param geneReference a reference (i.e., id generated) to find a gene in this history, which always refers to 'root' gene in the [evi]
-     *   @param archiveMutator mutate genes using archive-based methods if the method is enabled or supports this type of [this] gene.
-     */
-    open fun archiveMutation(randomness: Randomness,
-                             allGenes: List<Gene>,
-                             apc: AdaptiveParameterControl,
-                             selection: GeneMutationSelectionMethod,
-                             impact: GeneImpact?,
-                             geneReference: String,
-                             archiveMutator: ArchiveMutator,
-                             evi: EvaluatedIndividual<*>,
-                             targets: Set<Int>){
-        standardMutation(randomness, apc, allGenes)
+     * mutated gene should pass the check if needed, eg, DateGene
+     */
+    open fun mutationCheck() : Boolean = true
```

Comment:
I am confused about this one. can you provide more info, eg what is special in DateGene, and what is going to happen when this is false

Reply:
Regarding DateGene, I found that there is a condition check as below. 
```
            do {
                val gene = randomness.choose(listOf(year, month, day))
                gene.standardMutation(randomness, apc, allGenes)
            } while (onlyValidDates && !isValidDate())
```
If this is false (i.e., the mutated internal gene does not satisfy some constraints of the gene), the internal gene will be mutated again.

Reply:
ok, true. but let's add something like: "In some cases, we must have genes with 'valid' values. For example, a date with month 42 would be invalid. On the one hand, it can still be useful for robustness testing to provide such invalid values in a HTTP call. On the other hand, it would be pointless to try to add it directly into a database, as that SQL command would simply fail without any SUT code involved."